lib/auth/gitserver/gitserverv1/github.go

```diff
@@ -0,0 +1,133 @@
+	if in.Request == nil {
+		return nil, trace.BadParameter("missing github auth request")
+	}
+	if err := types.ValidateGitHubOrganizationName(in.Organization); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	authCtx, gitServer, err := s.authAndFindServerByOrg(ctx, in.Organization)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	s.cfg.Log.DebugContext(ctx, "Creating GitHub auth request for authenticated user.",
+		"user", authCtx.User.GetName(),
+		"org", gitServer.GetGitHub().Organization,
+		"integration", gitServer.GetGitHub().Integration,
+	)
+
+	// Make a "temporary" connector spec and save it in the request.
+	uuid := uuid.NewString()
+	spec, err := s.makeGithubConnectorSpec(ctx, uuid, gitServer.GetGitHub().Organization, gitServer.GetGitHub().Integration)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	in.Request.ConnectorID = uuid
+	in.Request.ConnectorSpec = spec
+	in.Request.AuthenticatedUser = authCtx.User.GetName()
+	in.Request.CertTTL = authCtx.Identity.GetIdentity().Expires.Sub(s.cfg.clock.Now())
+	in.Request.ClientLoginIP = authCtx.Identity.GetIdentity().LoginIP
+
+	// More params of in.Request will get updated and checked by
+	// s.cfg.GitHubAuthRequestCreator.
+	request, err := s.cfg.GitHubAuthRequestCreator(ctx, *in.Request)
+	return request, trace.Wrap(err)
+}
+
+func (s *Service) authAndFindServerByOrg(ctx context.Context, org string) (*authz.Context, types.Server, error) {
+	authCtx, err := s.authorize(ctx, types.VerbRead, types.VerbList)
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+
+	// We assume the list of servers is small and this function is called
+	// rarely. Use a cache when these assumptions change.
+	var servers []types.Server
+	var next string
+	for {
+		servers, next, err = s.cfg.Backend.ListGitServers(ctx, apidefaults.DefaultChunkSize, next)
+		if err != nil {
+			return nil, nil, trace.Wrap(err)
+		}
+		for _, server := range servers {
+			if spec := server.GetGitHub(); spec != nil && spec.Organization == org {
+				if err := s.checkAccess(authCtx, server); err == nil {
+					return authCtx, server, nil
+				}
+			}
+		}
+		if next == "" {
+			break
+		}
+	}
+	return nil, nil, trace.NotFound("git server with organization %q not found", org)
+}
+
+func (s *Service) makeGithubConnectorSpec(ctx context.Context, uuid, org, integration string) (*types.GithubConnectorSpecV3, error) {
+	ref, err := credentials.GetIntegrationRef(ctx, integration, s.cfg.Backend)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	cred, err := credentials.GetByPurpose(ctx, ref, credentials.PurposeGitHubOAuth, s.cfg.Backend)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	clientID, clientSecret := cred.GetOAuthClientSecret()
+	if clientID == "" || clientSecret == "" {
+		return nil, trace.BadParameter("no OAuth client ID or secret found for integration %v", integration)
+	}
+
+	return &types.GithubConnectorSpecV3{
```

Comment:
Just wanted to confirm - will the OAuth flow work seamlessly if GitHub is configured as an IdP for Teleport, while there is a second GitHub connector being used for IdP SSO logic?

Reply:
`createGithubUser` makes GitHub SSO users have the GitHub identity by default so the initial SSH user cert should have the GitHub identity without the help of this extra OAuth flow.